File: packages/ts-host/src/output-modes.ts
// Content-negotiation вывода (host-only) — ДВЕ НЕЗАВИСИМЫЕ ОСИ (A2A + AG-UI + A2UI), SSOT.
//
//  1. Формат текста — A2A `acceptedOutputModes` (media-типы). Аналог HTTP `Accept` (SHOULD).
//  2. Каталог(и) UI — A2UI-нативно: клиент шлёт `a2uiClientCapabilities.v0.9.supportedCatalogIds`
//     (список URL каталогов, по убыванию предпочтения) в метаданных КАЖДОГО сообщения; агент берёт
//     согласованное ПОДМНОЖЕСТВО (пересечение со своим списком); нет пересечения → UI не шлётся.
//
// Текст НЕ обязателен (AG-UI `content` опционален; A2A не требует текстовый part). Enforcement —
// здесь и в хендлере хоста. MIME-вокабуляр (`OUTPUT_MODE_*`) — agent-facing, живёт в `@ai37/agent-sdk`.
//
// См. docs/ecosystem/v2/10-agui-protocol.md (РЕШЕНИЕ 10), client_capabilities.json (@a2ui/web_core).

import { OUTPUT_MODE_TEXT, TEXT_OUTPUT_MODES, isTextOutputMode } from '@ai37/agent-sdk'

// ── Ось 1: формат текста (media-типы) ──────────────────────────────────────────────────────────

/**
 * Выбор текстового формата: первый текстовый mode из пересечения (client ∩ agentSupported) по
 * порядку клиента; если пересечения нет — `text/plain`. Текст всегда имеет валидный формат — но это
 * НЕ значит, что текст обязан эмититься (это решает агент, отдавая `message` или нет).
 */
export function negotiateText(
  accepted: readonly string[] | undefined,
  agentSupported: readonly string[],
): string {
  const supported = new Set(agentSupported)
  const list = Array.isArray(accepted) ? accepted : []
  return list.find((m) => isTextOutputMode(m) && supported.has(m)) ?? OUTPUT_MODE_TEXT
}

// ── Ось 2: каталог(и) UI (A2UI supportedCatalogIds) ─────────────────────────────────────────────

/** Версия A2UI-протокола в конверте capabilities. */
export const A2UI_CAPABILITIES_VERSION = 'v0.9'

/** Форма `a2uiClientCapabilities` (A2A message metadata / AG-UI forwardedProps). */
export interface A2uiClientCapabilities {
  [version: string]: { supportedCatalogIds?: string[] } | undefined
}

/**
 * Достаёт `supportedCatalogIds` из объекта-носителя (A2A `message.metadata` или AG-UI
 * `forwardedProps`), который может содержать `a2uiClientCapabilities.v0.9.supportedCatalogIds`.
 * Возвращает упорядоченный список URL каталогов (или []).
 */
export function readClientCapabilities(source: unknown): string[] {
  const caps = (source as { a2uiClientCapabilities?: A2uiClientCapabilities } | undefined)
    ?.a2uiClientCapabilities
  const ids = caps?.[A2UI_CAPABILITIES_VERSION]?.supportedCatalogIds
  return Array.isArray(ids) ? ids.filter((s): s is string => typeof s === 'string') : []
}

/** Поддерживает ли клиент данный каталог. */
export function clientSupportsCatalog(
  supportedCatalogIds: readonly string[] | undefined,
  agentCatalogId: string | undefined,
): boolean {
  return (
    !!agentCatalogId &&
    Array.isArray(supportedCatalogIds) &&
    supportedCatalogIds.includes(agentCatalogId)
  )
}

/**
 * Согласованное МНОЖЕСТВО каталогов: пересечение (client ∩ agent) в порядке предпочтения клиента.
 * Агент может эмитить несколько каталогов (напр. ai37-надмножество + базовый A2UI для деградации);
 * клиент рендерит то, что заявил в `supportedCatalogIds`. Пустой результат → UI не слать.
 */
export function negotiateCatalogs(
  supportedCatalogIds: readonly string[] | undefined,
  agentCatalogIds: string | readonly string[] | undefined,
): string[] {
  const agentSet = new Set(
    typeof agentCatalogIds === 'string'
      ? [agentCatalogIds]
      : Array.isArray(agentCatalogIds)
        ? agentCatalogIds
        : [],
  )
  if (agentSet.size === 0) return []
  const clientList = Array.isArray(supportedCatalogIds) ? supportedCatalogIds : []
  return clientList.filter((id) => agentSet.has(id))
}

/**
 * Скалярный выбор каталога (обратная совместимость): первый из согласованного множества, либо `null`.
 */
export function negotiateCatalog(
  supportedCatalogIds: readonly string[] | undefined,
  agentCatalogIds: string | readonly string[] | undefined,
): string | null {
  return negotiateCatalogs(supportedCatalogIds, agentCatalogIds)[0] ?? null
}

// ── Сводная негоциация ──────────────────────────────────────────────────────────────────────────

/**
 * Резолвнутая негоциация вывода (две оси):
 * - `text` — формат текста, ЕСЛИ агент его эмитит (текст не обязателен).
 * - `catalogIds` — согласованные каталоги A2UI (порядок предпочтения клиента); пусто → UI не слать.
 * - `catalogId` — первый из `catalogIds` (или `null`); скалярный alias для обратной совместимости.
 */
export interface OutputNegotiation {
  text: string
  catalogIds: string[]
  catalogId: string | null
}

export interface NegotiateOutputArgs {
  /** A2A `acceptedOutputModes` клиента (media-типы текста). */
  acceptedOutputModes?: readonly string[]
  /** Текстовые форматы, которые умеет агент (agent-card `defaultOutputModes`). */
  agentTextModes?: readonly string[]
  /** Каталоги, заявленные клиентом (`a2uiClientCapabilities.supportedCatalogIds`). */
  supportedCatalogIds?: readonly string[]
  /** Каталог(и), которые эмитит агент (один — `CATALOG_ID`, или массив с base для деградации). */
  agentCatalogIds?: string | readonly string[]
}

export function negotiateOutput(args: NegotiateOutputArgs): OutputNegotiation {
  const catalogIds = negotiateCatalogs(args.supportedCatalogIds, args.agentCatalogIds)
  return {
    text: negotiateText(args.acceptedOutputModes, args.agentTextModes ?? TEXT_OUTPUT_MODES),
    catalogIds,
    catalogId: catalogIds[0] ?? null,
  }
}

/**
 * Отсекает A2UI-компоненты, если каталог не согласован (бинарный enforcement-хелпер).
 * Каталог согласован (`catalogIds` непуст) → компоненты как есть; иначе → [].
 */
export function filterA2uiComponents<T>(
  components: readonly T[] | undefined,
  negotiation: OutputNegotiation,
): T[] {
  if (negotiation.catalogIds.length === 0) return []
  return components ? [...components] : []
}

/**
 * Per-component enforcement: оставляет только компоненты, чей каталог в согласованном множестве.
 * Компонент без `catalogId` относится к первичному каталогу (`catalogIds[0]`). Это включает
 * роутинг A (соседние сообщения из разных каталогов) и деградацию на base.
 */
export function filterA2uiByCatalog<T extends { catalogId?: string }>(
  components: readonly T[] | undefined,
  negotiation: OutputNegotiation,
): T[] {
  if (!components || negotiation.catalogIds.length === 0) return []
  const allowed = new Set(negotiation.catalogIds)
  const primary = negotiation.catalogIds[0]
  return components.filter((c) => allowed.has(c.catalogId ?? primary))
}

File: packages/python-host/src/ai37_agent_host/output_modes.py
"""Content-negotiation вывода (host-only) — порт ``ts-host/src/output-modes.ts``.

Две независимые оси:
  1. Формат текста — A2A ``acceptedOutputModes`` (media-типы). Аналог HTTP ``Accept``.
  2. Каталог(и) UI — A2UI-нативно: ``a2uiClientCapabilities.v0.9.supportedCatalogIds``.

MIME-вокабуляр (``OUTPUT_MODE_*``) — agent-facing, живёт в ``ai37-agent-sdk``.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any

from ai37_agent_sdk import OUTPUT_MODE_TEXT, TEXT_OUTPUT_MODES, is_text_output_mode

from .types import A2uiComponent, OutputNegotiation

#: Версия A2UI-протокола в конверте capabilities.
A2UI_CAPABILITIES_VERSION = "v0.9"


# ── Ось 1: формат текста (media-типы) ────────────────────────────────────────


def negotiate_text(
    accepted: Sequence[str] | None,
    agent_supported: Sequence[str],
) -> str:
    """Первый текстовый mode из client∩agent по порядку клиента, иначе OUTPUT_MODE_TEXT."""
    supported = set(agent_supported)
    for mode in accepted or []:
        if is_text_output_mode(mode) and mode in supported:
            return mode
    return OUTPUT_MODE_TEXT


# ── Ось 2: каталог(и) UI (A2UI supportedCatalogIds) ──────────────────────────


def read_client_capabilities(source: Any) -> list[str]:
    """Упорядоченный supportedCatalogIds из носителя (A2A metadata / AG-UI forwardedProps)."""
    caps = source.get("a2uiClientCapabilities") if isinstance(source, dict) else None
    version = caps.get(A2UI_CAPABILITIES_VERSION) if isinstance(caps, dict) else None
    ids = version.get("supportedCatalogIds") if isinstance(version, dict) else None
    if isinstance(ids, list):
        return [s for s in ids if isinstance(s, str)]
    return []


def client_supports_catalog(
    supported_catalog_ids: Sequence[str] | None,
    agent_catalog_id: str | None,
) -> bool:
    return bool(agent_catalog_id) and isinstance(supported_catalog_ids, list | tuple) and (
        agent_catalog_id in supported_catalog_ids
    )


def negotiate_catalogs(
    supported_catalog_ids: Sequence[str] | None,
    agent_catalog_ids: str | Sequence[str] | None,
) -> list[str]:
    """Пересечение (client ∩ agent) в порядке предпочтения клиента. Пусто → UI не слать."""
    if isinstance(agent_catalog_ids, str):
        agent_set = {agent_catalog_ids}
    elif isinstance(agent_catalog_ids, list | tuple):
        agent_set = set(agent_catalog_ids)
    else:
        agent_set = set()
    if not agent_set:
        return []
    client_list = supported_catalog_ids if isinstance(supported_catalog_ids, list | tuple) else []
    return [cid for cid in client_list if cid in agent_set]


def negotiate_catalog(
    supported_catalog_ids: Sequence[str] | None,
    agent_catalog_ids: str | Sequence[str] | None,
) -> str | None:
    """Скалярный выбор каталога: первый из согласованного множества, либо None."""
    catalogs = negotiate_catalogs(supported_catalog_ids, agent_catalog_ids)
    return catalogs[0] if catalogs else None


# ── Сводная негоциация ───────────────────────────────────────────────────────


def negotiate_output(
    *,
    accepted_output_modes: Sequence[str] | None = None,
    agent_text_modes: Sequence[str] | None = None,
    supported_catalog_ids: Sequence[str] | None = None,
    agent_catalog_ids: str | Sequence[str] | None = None,
) -> OutputNegotiation:
    catalog_ids = negotiate_catalogs(supported_catalog_ids, agent_catalog_ids)
    return OutputNegotiation(
        text=negotiate_text(
            accepted_output_modes,
            agent_text_modes if agent_text_modes is not None else TEXT_OUTPUT_MODES,
        ),
        catalog_ids=catalog_ids,
        catalog_id=catalog_ids[0] if catalog_ids else None,
    )


def filter_a2ui_components(
    components: Sequence[A2uiComponent] | None,
    negotiation: OutputNegotiation,
) -> list[A2uiComponent]:
    """Бинарный enforcement: каталог согласован → компоненты как есть; иначе → []."""
    if not negotiation.catalog_ids:
        return []
    return list(components) if components else []


def filter_a2ui_by_catalog(
    components: Sequence[A2uiComponent] | None,
    negotiation: OutputNegotiation,
) -> list[A2uiComponent]:
    """Per-component enforcement: оставляет компоненты, чей каталог в согласованном множестве.

    Компонент без ``catalog_id`` относится к первичному каталогу (``catalog_ids[0]``).
    """
    if not components or not negotiation.catalog_ids:
        return []
    allowed = set(negotiation.catalog_ids)
    primary = negotiation.catalog_ids[0]
    return [c for c in components if (c.catalog_id or primary) in allowed]
